### ISSUE TITLE
Hellforge accepts input into the output slot, and extraction can pull from input slots as well as output.

### ISSUE THREAD

**Issue by mopsyd** [WayofTime/BloodMagic#2059]
#### Issue Description:
The hellforge has apparently no understanding anymore of which of its slots should be accessible for input/output by other machines/pipes/hoppers/etc. This makes automating them impossible, which cascades into automating will production also impossible because it requires catalysts made in the hellforge of a volume significantly higher than is practical to manually babysit.

Extracting from input slots (and the orb slot) also applies to the alchemy bench, although I was able to work around this with filters.

#### What happens:
AE system, functional storage, and ender io conduits all insert readily into the output slot of the hellforge if the other four slots are full already. This halts the machine from working and there does not appear to be any meaningful way to stop them from doing this.

#### What you expected to happen:
Other blocks should not be able to insert into output or extract from input slots/orb slots/etc (only from output slots).

#### Steps to reproduce:

1. Place a hellfire forge
2. Pipe items into it with applied energistics, the push module from functional storage, or ender io conduits (these are the ones I have tested, likely others as well)
3. Wait for it to fill the stack limit. Subsequent items will start filling the output slot.
...
4. You can also readily see it extracting from the wrong slots by reversing the direction of the pipes/conduits/bus/etc

____
#### Affected Versions (Do *not* use "latest"):

- BloodMagic: 3.3.3-45
- Minecraft: 1.20.1
- Forge: 47.2.20

**Comment by wrincewind** [comment]
as a temporary workaround, most auto-crafting systems have a 'wait until craft is complete before inserting ingredients for next craft'. I believe Applied Energistics crafters do have this. I know it's not a perfect solution but it should fix the problem until we can get an update pushed out.

**Comment by mopsyd** [comment]
Yea I use only my AE system in blocking mode and that seems to work relatively ok, however every other type of pipe/conduit/etc fails. I am just reporting for consistency's sake, I already have a workaround but ideally the expected behavior would be consistent with vanilla and most every other mod machine.

**Comment by mopsyd** [comment]
As an aside, there are two theoretical approaches that I think would be consistent with the flavor of the mod and also facilitate automation compatibility. The easiest would be a button to lock the recipe, which would force items piped in to only fill the corresponding slot(s). The other way I have seen it done is akin to the thermal series or mekanism where you can control the specific i/o of each slot, but that would probably be considerably more work creating a ui for it. Just using the vanilla logic would also work fine if the filtering was done externally.

**Comment by Faultedskink** [comment]
i notice this also happens with blood magic's routing node system. unless the input for the hellfire forge is very slow and perfectly distributed to the 4 input slots, its just not possible to automate a hellfire forge PERIOD. automated outputting is definitely still possible and easy, but not inputting, and at the end of the day its still babysitting will automation which is the entire thing were trying to accomplish.

**Comment by stellanera98** [comment]
If you set the output filter to input less than a stack for each ingredient it works fine.

As for will automation there is nothing to babysit there once you have it set up
You need 2 chunks, one of them to farm the will the other for the forge. In the farming chunk you set up a Crack of the Fractured Crystal ritual and in its operating area you plop down as many crystals as you want. If you only want it for crafting just raw will crystals are fine. Next you want to set up a Call of the Zephyr to collect the split off crystal items (or a Hopperhock, Item Collector, Vacuum Hopper, etc) and attach an input node to the inventory the collected items go with an empty blacklist filter. Connect it to a master routing node and the master routing node to an output node. Around the output node place hoppers pointing into a Demon Crucible, one hopper/Crucible pair for each will type you are trying to farm. Set up whitelist filters for each will type and limit them to at least the same amount as you have crystals of that type to insure that from whats left at least one more can grow. Using more makes it safer but it takes longer until you get will to use. Set the priority for these ones to at least 1.
Now you want to place another output node and crucible (hopper optional but recommended) in the crafting chunk where you have your hellfire forge with a tartaric gem (preferably maxed and filled) inside of it. The output node gets a whitelist filter for the same type the tartaric gem accepts and you can leave the number as is so it'll just fill any overflow to it. Leave that nodes priority at 0. If this priority is higher than the ones from the farming chunk they wont get any will and the farm doesnt work anymore.

You can add more forges to the crafting chunk and you can set up a Gathering of the Forsaken Souls ritual (covering the crystal area) to speed up the process but that requires some way of automating the production of the catalysts which is currently not possible in vanilla blood magic

**Comment by Faultedskink** [comment]
yes, we're trying to get the gathering of the forsaken souls ritual automated. of corse will production can be automated with a 10 will profit with the CotFC ritual as it is. as a side note, I've found the output nodes can input directly into the demon crucibles, which enables one to set up a demon crucible pyramid with the output node in the middle with filters going to each crucible set to burn their respective aspects, and it all fits in the same chunk.

to attempt to automate the GotFS ritual I've got this set up:
-farm produces the ingredients for each catalyst (tau fruit, nether wart, pumpkin, melon, beetroot, wheat and potato; with a skeleton mob spawner/grinder for bone, and cobblestone generator with a create mechanical drill for cobblestone), Reap of the Harvest moon and Ritual of the Green Grove to speed up and harvest all products with a CotZ ritual to pick up all drops.
-bones, pumpkins and melons get routed to "Rhythm of the Beating Anvil" rituals each to produce bone meal, pumpkin seeds and melon seeds
-bone meal and tau fruit get routed to one alchemy table to produce tau oil, while cobblestone gets routed to another alchemy table with a lava sigil to produce sulfur. (this part is easy because the alchemy table has an interface that for each input and output slot to be accessed by any cardinal direction around the table if at all. ill get back to this.)
-5 hellfire forges with output nodes connected input the ingredients for each desired catalyst. **here is where the problem arises because all of the slots seem to think of themselves as input slots for ALL ingredients. for instance when tau oil runs out, but the pumpkin seeds are still full and there's a backlog of pumpkin seeds ready to be put into the forge, it'll fill in the newly empty slot and the forge wont accept tau oil anymore because there's no vacant slot. furthermore, when i said all slots, that included the output slot, as even when all ingredients are in place and the tartaric gem holds will, the forge will start crafting the catalyst, only to get cut off by pumpkin seeds plopping themselves into the output slot for no apparent reason. here it would be nice to have something like what the alchemy table has where each slot can be configured to accept items from only one direction.**
-the crafted catalysts are then routed to the chest for the GotFS ritual to use.

as you can see in the bold text, in step 4, this is not really possible to automate catalyst production effectively. which is why i suggested an interface with similar functionality to the alchemy table: that is to say each cardinal direction can be directed to an input slot to accept an item, but only ever fill that slot.

sorry for the over explanation, i just want to make this rather infuriating issue as clear as possible so a solution can be developed

**Comment by stellanera98** [comment]
Unfortunately WayofTime is currently on a hiatus so its very unlikely you'll get any additions anytime soon, though I agree that there should be improvements made to the way the forge handles its sides/slots.

I said it wasnt possible with just BM to automate the catalysts because I was under the impression that the harvesting ritual did not harvest nether warts, but its been a while since I messed with it so I might be misremembering that.

Since you seem to already have one forge set up for each catalyst you could use beating anvil rituals for those.
Im also pretty sure if you limit the amount of items put into the forges they cannot overfill the slots. So if you adjust the filters to at max put [the respective max stack size] in the forge it could never start a second stack of any one ingredient (and since there are only 4 for any given forge recipe, never touch the output slot) even if you run out of one completely.

I am aware that the output nodes can put will directly into the crucibles, the reason I put hoppers in between is so that there is always enough backlog of will items to produce one more, even if you were to eat through all the stored will at once. The hoppers there are technically optional, I just like to be on the save side

**Comment by Faultedskink** [comment]
ill need to have a fiddle with the different filter types. i was under the impression that filters couldnt output full stacks, unless this is something from another mod that im unaware of and havent plugged into my pack

**Comment by stellanera98** [comment]
Im talking about regular Item Filters from Blood Magic.
When you just set the items desired it will take any it can (when used in input) and place as much as it can (output)
When you set a number as well it will leave that amount in the input inventory (when used in input) and place only up to that amount into the output inventory (when used in output)

So if you set all your catalyst ingredients to at most be 64 (I dont think any of them stack to lower) it will only fill them up to 64 items of that type in the forge. So if there are 0 Pumpkin seeds it will try to put 64 of them (16 if you dont have any upgrades in the master node) and if theres 63 already in the forge it will only try to put one more to make 64
Theres also a tooltip denoting the amounts or "Any" if there was no limit set.
Theres also a blacklist option and Im not entirely sure how the limit interacts there but in this case the whitelist seems easier to use anyways.

I think all the other filter types have this limit amount as well but I havent messed with them in a long time so not 100% sure

**Comment by Faultedskink** [comment]
ahh, yes, ive done some more fiddling around and found this works as desired. its a bit unclear and indirect to get the outcome, but it works very well. thank you
